tests/capi/unittest_capi_service_training_offloading.cc

```diff
@@ -0,0 +1,292 @@
+/**
+ * @file        unittest_capi_offloading_service.cc
+ * @date        5 April 2024
+ * @brief       Unit test for ML Service C-API offloading service.
+ * @see         https://github.com/nnstreamer/api
+ * @author      Hyunil Park <hyunil46.park@samsung.com>
+ * @bug         No known bugs
+ */
+
+#include <gtest/gtest.h>
+#include <gio/gio.h>
+#include <glib/gstdio.h>
+#include <ml-api-inference-pipeline-internal.h>
+#include <ml-api-internal.h>
+#include <ml-api-service-private.h>
+#include <ml-api-service.h>
+
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+
+#include "ml-api-service-offloading.h"
+
+/**
+ * @brief Internal function to get the config file path.
+ */
+static gchar *
+_get_config_path (const gchar *config_name)
+{
+  const gchar *root_path = g_getenv ("MLAPI_SOURCE_ROOT_PATH");
+
+  /* Supposed to run test in build directory. */
+  if (root_path == NULL)
+    root_path = "..";
+
+  gchar *config_file = g_build_filename (
+      root_path, "tests", "test_models", "config", config_name, NULL);
+
+  return config_file;
+}
+
+/**
+ * @brief Test base class for Database of ML Service API.
+ */
+class MLServiceTrainingOffloading : public ::testing::Test
+{
+  protected:
+  GTestDBus *dbus;
+  GBusType bus_type;
+  int status;
+  ml_service_h client_h;
+  ml_service_h server_h;
+
+  public:
+  /**
+   * @brief Setup method for each test case.
+   */
+  void SetUp () override
+  {
+    g_autofree gchar *services_dir = g_build_filename ("/usr/bin/ml-test/services", NULL);
+
+    dbus = g_test_dbus_new (G_TEST_DBUS_NONE);
+    ASSERT_NE (nullptr, dbus);
+
+    g_test_dbus_add_service_dir (dbus, services_dir);
+
+    g_test_dbus_up (dbus);
+#if defined(ENABLE_GCOV)
+    bus_type = G_BUS_TYPE_SYSTEM;
+#else
+    bus_type = G_BUS_TYPE_SESSION;
+#endif
+
+    /* should start receiver */
+    g_autofree gchar *receiver_config
+        = _get_config_path ("training_offloading_receiver.conf");
+    status = ml_service_new (receiver_config, &client_h);
+    ASSERT_EQ (status, ML_ERROR_NONE);
+
+    g_autofree gchar *sender_config = _get_config_path ("training_offloading_sender.conf");
+    status = ml_service_new (sender_config, &server_h);
+    ASSERT_EQ (status, ML_ERROR_NONE);
+  }
+
+  /**
+   * @brief Teardown method for each test case.
+   */
+  void TearDown () override
+  {
+    if (dbus) {
+      g_test_dbus_down (dbus);
+      g_object_unref (dbus);
+    }
+
+    status = ml_service_destroy (client_h);
+    EXPECT_EQ (ML_ERROR_NONE, status);
+
+    status = ml_service_destroy (server_h);
+    EXPECT_EQ (ML_ERROR_NONE, status);
+  }
+
+  /**
+   * @brief Get available port number.
+   */
+  static guint _get_available_port (void)
```

Comment:
The same function is implemented in three sources. Let's merge into one to prevent code duplication.
Let's do it later.